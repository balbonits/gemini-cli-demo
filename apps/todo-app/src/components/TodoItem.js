class TodoItem extends HTMLElement {
    constructor() {
        super();
        this.attachShadow({ mode: 'open' });
    }

    set todo(todo) {
        this._todo = todo;
        this.render();
    }

    get todo() {
        return this._todo;
    }

    render() {
        this.shadowRoot.innerHTML = `
            <style>
                li {
                    display: flex;
                    align-items: center;
                    padding: 0.5rem 0;
                    border-bottom: 1px solid #ccc;
                }
                .completed {
                    text-decoration: line-through;
                    color: #aaa;
                }
                input[type="checkbox"] {
                    margin-right: 0.5rem;
                }
                .edit-input {
                    flex: 1;
                    padding: 0.5rem;
                    border: 1px solid #ccc;
                    border-radius: 4px;
                }
                button {
                    margin-left: 0.5rem;
                    background: none;
                    border: none;
                    cursor: pointer;
                }
                .delete-btn {
                    color: red;
                }
                .edit-btn {
                    color: #007bff;
                }
                .save-btn {
                    color: green;
                }
                @media (max-width: 400px) {
                    button {
                        padding: 0.5rem;
                        margin-left: 0.5rem;
                    }
                }
            </style>
            <li class="${this._todo.completed ? 'completed' : ''}">
                <input type="checkbox" ${this._todo.completed ? 'checked' : ''}>
                <span class="todo-text">${this._todo.text}</span>
                <input type="text" class="edit-input" style="display: none;" />
                <button class="edit-btn">Edit</button>
                <button class="save-btn" style="display: none;">Save</button>
                <button class="delete-btn">Delete</button>
            </li>
        `;

        this.shadowRoot.querySelector('input[type="checkbox"]').addEventListener('change', () => {
            this.dispatchEvent(new CustomEvent('toggle-complete', { detail: this._todo }));
        });

        this.shadowRoot.querySelector('.delete-btn').addEventListener('click', () => {
            this.dispatchEvent(new CustomEvent('delete-todo', { detail: this._todo }));
        });

        this.shadowRoot.querySelector('.edit-btn').addEventListener('click', () => {
            this.enterEditMode();
        });

        this.shadowRoot.querySelector('.save-btn').addEventListener('click', () => {
            this.saveTodo();
        });
    }

    enterEditMode() {
        const span = this.shadowRoot.querySelector('.todo-text');
        const input = this.shadowRoot.querySelector('.edit-input');
        const editBtn = this.shadowRoot.querySelector('.edit-btn');
        const saveBtn = this.shadowRoot.querySelector('.save-btn');

        span.style.display = 'none';
        input.style.display = 'block';
        input.value = this._todo.text;
        editBtn.style.display = 'none';
        saveBtn.style.display = 'block';
        input.focus();
    }

    saveTodo() {
        const input = this.shadowRoot.querySelector('.edit-input');
        const newText = input.value.trim();
        if (newText && newText !== this._todo.text) {
            this.dispatchEvent(new CustomEvent('edit-todo', { detail: { oldTodo: this._todo, newText } }));
        }
        this.exitEditMode();
    }

    exitEditMode() {
        const span = this.shadowRoot.querySelector('.todo-text');
        const input = this.shadowRoot.querySelector('.edit-input');
        const editBtn = this.shadowRoot.querySelector('.edit-btn');
        const saveBtn = this.shadowRoot.querySelector('.save-btn');

        span.style.display = 'block';
        input.style.display = 'none';
        editBtn.style.display = 'block';
        saveBtn.style.display = 'none';
    }
}

customElements.define('todo-item', TodoItem);