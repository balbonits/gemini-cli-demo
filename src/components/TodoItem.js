class TodoItem extends HTMLElement {
    constructor() {
        super();
        this.attachShadow({ mode: 'open' });
    }

    set todo(todo) {
        this._todo = todo;
        this.render();
    }

    get todo() {
        return this._todo;
    }

    render() {
        this.shadowRoot.innerHTML = `
            <style>
                li {
                    display: flex;
                    align-items: center;
                    padding: 0.5rem 0;
                    border-bottom: 1px solid #ccc;
                }
                .completed {
                    text-decoration: line-through;
                    color: #aaa;
                }
                input[type="checkbox"] {
                    margin-right: 0.5rem;
                }
                button {
                    margin-left: auto;
                    background: none;
                    border: none;
                    color: red;
                    cursor: pointer;
                }
            </style>
            <li class="${this._todo.completed ? 'completed' : ''}">
                <input type="checkbox" ${this._todo.completed ? 'checked' : ''}>
                <span>${this._todo.text}</span>
                <button>Delete</button>
            </li>
        `;

        this.shadowRoot.querySelector('input[type="checkbox"]').addEventListener('change', () => {
            this.dispatchEvent(new CustomEvent('toggle-complete', { detail: this._todo }));
        });

        this.shadowRoot.querySelector('button').addEventListener('click', () => {
            this.dispatchEvent(new CustomEvent('delete-todo', { detail: this._todo }));
        });
    }
}

customElements.define('todo-item', TodoItem);