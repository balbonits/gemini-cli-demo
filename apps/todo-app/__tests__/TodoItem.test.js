
import '../src/components/TodoItem.js';

describe('TodoItem', () => {
  let todoItem;
  const todo = { text: 'Test Todo', completed: false };

  beforeEach(async () => {
    document.body.innerHTML = '<todo-item></todo-item>';
    await customElements.whenDefined('todo-item');
    todoItem = document.querySelector('todo-item');
    todoItem.todo = todo;
  });

  afterEach(() => {
    document.body.innerHTML = '';
  });

  test('should render correctly', () => {
    expect(todoItem).toMatchSnapshot();
  });

  test('should dispatch toggle-complete event on checkbox change', () => {
    const eventSpy = jest.fn();
    todoItem.addEventListener('toggle-complete', eventSpy);
    todoItem.shadowRoot.querySelector('input[type="checkbox"]').click();
    expect(eventSpy).toHaveBeenCalled();
  });

  test('should dispatch delete-todo event on delete button click', () => {
    const eventSpy = jest.fn();
    todoItem.addEventListener('delete-todo', eventSpy);
    todoItem.shadowRoot.querySelector('.delete-btn').click();
    expect(eventSpy).toHaveBeenCalled();
  });

  test('should enter edit mode on edit button click', () => {
    todoItem.shadowRoot.querySelector('.edit-btn').click();
    expect(todoItem.shadowRoot.querySelector('.edit-input').style.display).toBe('block');
    expect(todoItem.shadowRoot.querySelector('.save-btn').style.display).toBe('block');
  });

  test('should dispatch edit-todo event on save button click', () => {
    const eventSpy = jest.fn();
    todoItem.addEventListener('edit-todo', eventSpy);
    todoItem.shadowRoot.querySelector('.edit-btn').click();
    const editInput = todoItem.shadowRoot.querySelector('.edit-input');
    editInput.value = 'Updated Todo';
    todoItem.shadowRoot.querySelector('.save-btn').click();
    expect(eventSpy).toHaveBeenCalledWith(new CustomEvent('edit-todo', { detail: { oldTodo: todo, newText: 'Updated Todo' } }));
  });
});
