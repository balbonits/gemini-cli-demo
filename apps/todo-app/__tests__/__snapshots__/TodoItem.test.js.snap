// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TodoItem should render correctly 1`] = `<todo-item />`;
